From a9413467924080c730adec3353a69c4e5e6024fa Mon Sep 17 00:00:00 2001
From: Guido Ostkamp <Guido.Ostkamp@partner.bmw.de>
Date: Fri, 14 Feb 2020 15:58:51 +0100
Subject: Introduce mosquitto_connect_bind_async_v5() API

The mosquitto_connect_bind_async_v5() API extends the
mosquitto_connect_bind_async() API by allowing to specify
MQTT v5 properties for CONNECT.
---
 lib/connect.c      | 16 ++++++++++++++++
 lib/linker.version |  1 +
 lib/mosquitto.h    | 42 ++++++++++++++++++++++++++++++++++++++++++
 3 files changed, 59 insertions(+)

diff --git a/lib/connect.c b/lib/connect.c
index fd5c497..882206c 100644
--- a/lib/connect.c
+++ b/lib/connect.c
@@ -140,6 +140,22 @@ int mosquitto_connect_bind_async(struct mosquitto *mosq, const char *host, int p
 }
 
 
+int mosquitto_connect_bind_async_v5(struct mosquitto *mosq, const char *host, int port, int keepalive, const char *bind_address, const mosquitto_property *properties)
+{
+	int rc;
+
+	if(properties){
+		rc = mosquitto_property_check_all(CMD_CONNECT, properties);
+		if(rc) return rc;
+	}
+
+	rc = mosquitto__connect_init(mosq, host, port, keepalive, bind_address);
+	if(rc) return rc;
+
+	return mosquitto__reconnect(mosq, false, properties);
+}
+
+
 int mosquitto_reconnect_async(struct mosquitto *mosq)
 {
 	return mosquitto__reconnect(mosq, false, NULL);
diff --git a/lib/linker.version b/lib/linker.version
index 017d5a5..3043ac5 100644
--- a/lib/linker.version
+++ b/lib/linker.version
@@ -94,6 +94,7 @@ MOSQ_1.5 {
 
 MOSQ_1.6 {
 	global:
+		mosquitto_connect_bind_async_v5;
 		mosquitto_connect_bind_v5;
 		mosquitto_connect_v5_callback_set;
 		mosquitto_disconnect_v5;
diff --git a/lib/mosquitto.h b/lib/mosquitto.h
index 97e3480..3e4dddc 100644
--- a/lib/mosquitto.h
+++ b/lib/mosquitto.h
@@ -581,6 +581,48 @@ libmosq_EXPORT int mosquitto_connect_async(struct mosquitto *mosq, const char *h
 libmosq_EXPORT int mosquitto_connect_bind_async(struct mosquitto *mosq, const char *host, int port, int keepalive, const char *bind_address);
 
 /*
+ * Function: mosquitto_connect_bind_async_v5
+ *
+ * Connect to an MQTT broker. This is a non-blocking call.
+ * If you use <mosquitto_connect_bind_async_v5> your client must use the
+ * threaded interface <mosquitto_loop_start>. If you need to use <mosquitto_loop>,
+ * you must use <mosquitto_connect> to connect the client.
+ *
+ * This extends the functionality of <mosquitto_connect_bind_async> by adding the
+ * MQTT v5 properties.
+ *
+ * Use e.g. <mosquitto_property_add_string> and similar to create a list of
+ * properties, then attach them to this publish. Properties need freeing with
+ * <mosquitto_property_free_all>.
+ *
+ * May be called before or after <mosquitto_loop_start>.
+ *
+ * Parameters:
+ * 	mosq -         a valid mosquitto instance.
+ * 	host -         the hostname or ip address of the broker to connect to.
+ * 	port -         the network port to connect to. Usually 1883.
+ * 	keepalive -    the number of seconds after which the broker should send a PING
+ *                 message to the client if no other messages have been exchanged
+ *                 in that time.
+ *  bind_address - the hostname or ip address of the local network interface to
+ *                 bind to.
+ *  properties -   the MQTT 5 properties for the connect (not for the Will).
+ *
+ * Returns:
+ * 	MOSQ_ERR_SUCCESS - on success.
+ * 	MOSQ_ERR_INVAL -   if the input parameters were invalid.
+ * 	MOSQ_ERR_ERRNO -   if a system call returned an error. The variable errno
+ *                     contains the error code, even on Windows.
+ *                     Use strerror_r() where available or FormatMessage() on
+ *                     Windows.
+ *
+ * See Also:
+ * 	<mosquitto_connect_async>, <mosquitto_connect>, <mosquitto_connect_bind>,
+ *  <mosquitto_connect_bind_async>
+ */
+libmosq_EXPORT int mosquitto_connect_bind_async_v5(struct mosquitto *mosq, const char *host, int port, int keepalive, const char *bind_address, const mosquitto_property *properties);
+
+/*
  * Function: mosquitto_connect_srv
  *
  * Connect to an MQTT broker.
-- 
2.7.4

